docs: add tail sampling docs and example (#3094)
kubeshop/tracetest: examples/quick-start-tail-sampling/README.md

> # Quick Start - Trace-based tests with Tail Sampling configuration

Reviewer: ```suggestion
# Quick Start - Trace-based Tests with Tail Sampling Configuration
```

> > [Read the detailed recipe for setting up OpenTelemetry Collector with Tractest in our documentation.](https://docs.tracetest.io/examples-tutorials/recipes/running-tracetest-without-a-trace-data-store)

Reviewer 2: ```suggestion
> [Read the detailed recipe for setting up OpenTelemetry Collector with Tracetest in our documentation.](https://docs.tracetest.io/examples-tutorials/recipes/running-tracetest-without-a-trace-data-store)
```

This is a simple quick start example on how to set up `tail_sampling` into OTel Collector, allowing Tracetest to run tests in environments where we have a probabilistic sampler enabled and a percentage of the traces are not sent to the final data store.

## Scenario

In this scenario, we have on [Go API](./simple-go-service/) that sends Trace data to one instance of [OTel Collector](https://opentelemetry.io/docs/collector/), which samples 50% of the traces and sends it to [Jaeger](https://www.jaegertracing.io/) data store.

```mermaid
---
title: Initial observability architecture
---
flowchart LR
    GoAPI["Go API"]
    OTelCol["OTel Collector"]
    DataStore[("Jaeger")]

    GoAPI -- Send traces --> OTelCol
    OTelCol -- Forward 50% of Traces --> DataStore
``````

The collector configuration is defined by a `traces/jaeger` pipeline with one receiver, two processors (one to send traces in batch and the other set up the probabilistic sampling) and one exporter to Jaeger:

```yaml
receivers:
  otlp:
    protocols:
      grpc:
      http:

processors:
  batch:
    timeout: 100ms
  probabilistic_sampler:
    hash_seed: 22
    sampling_percentage: 50.0 # 50%

exporters:
  otlp/jaeger:

>     endpoint: jaeger:4317 # Send traces to Jager

Reviewer 2: ```suggestion
    endpoint: jaeger:4317 # Send traces to Jaeger
```

    tls:
      insecure: true

service:
  pipelines:
    traces/jaeger:
      receivers: [otlp]
      processors: [batch, probabilistic_sampler]
      exporters: [otlp/jaeger]

```

> However, to run a Trace-based test on this scenario and validate if the API is ok, we need to have the testing trace collected by our OTelCollector, how can we do that?

Reviewer: ```suggestion
However, to run a Trace-based test on this scenario and validate if the API is ok, we need to have the testing trace collected by our OTelCollector. How can we do that?
```

> Since Tracetest always starts a trace by adding a `tracetest=true` key/value to [TraceState](https://opentelemetry.io/docs/specs/otel/trace/api/#tracestate), that is [propagated](https://opentelemetry.io/docs/instrumentation/js/propagation/), we can filter these traces on the collector and send them to Tracetest.

Reviewer 2: I would add a side note near this paragraph about the need for updated open telemetry libraries because `tracestate` is relatively new and not all languages SDKs were supporting it last year.

Example, the Javascript SDK only started supporting the `tracestate` field early this year.

Author: I have not added it because I want to check the SDK states and see the feedback from OTel team first.

> Since Tracetest always starts a trace by adding a `tracetest=true` key/value to [TraceState](https://opentelemetry.io/docs/specs/otel/trace/api/#tracestate), that is [propagated](https://opentelemetry.io/docs/instrumentation/js/propagation/), we can filter these traces on the collector and send them to Tracetest.

Reviewer: ```suggestion
Since Tracetest always starts a trace by adding a `tracetest=true` key/value to [TraceState](https://opentelemetry.io/docs/specs/otel/trace/api/#tracestate) that is [propagated](https://opentelemetry.io/docs/instrumentation/js/propagation/), we can filter these traces on the collector and send them to Tracetest.
```

> To do it we need to set up Tracetest to use an OTLP Data Store on its provision ([here](./tracetest/tracetest.provision.yaml)) and also set up another pipeline with a [Tail Sampling](https://opentelemetry.io/docs/concepts/sampling/#tail-sampling) processor on OTel Collector, filtering by TraceState `tracetest=true` value.

Reviewer: ```suggestion
To do this we need to set up Tracetest to use an OTLP Data Store on its provision ([here](./tracetest/tracetest.provision.yaml)) and also set up another pipeline with a [Tail Sampling](https://opentelemetry.io/docs/concepts/sampling/#tail-sampling) processor on the OTel Collector, filtering by TraceState `tracetest=true` value.
```

We can do that by adding the following configuration to OTel Collector (the entire collector config YAML is [here](./tracetest/collector.config.yaml)):
```yaml
#...

processors:
  #...
  tail_sampling:
    decision_wait: 10s
    num_traces: 100
    expected_new_traces_per_sec: 10
    policies:
      [
        {
          name: Accept only traces that started from tracetest,
          type: trace_state,
          trace_state: {
            key: tracetest,
            values: ["true"]
          }
        }
      ]

exporters:
  #...
  otlp/tracetest:
    endpoint: tracetest:4317 # Send traces to Tracetest.
    tls:
      insecure: true

service:
  pipelines:
    #...
    traces/tracetest:
      receivers: [otlp]
      processors: [batch, tail_sampling]

>       exporters: [otlp/tracetest, logging]

Reviewer 2: Would be nice to have a third configuration mixing both approaches, still doing probabilistic sampling, but making sure all tracetest traces go to a datastore (either jaeger or tracetest)

Author: No, this is why I'm using an OTel Collector config to do the tail sampling.

```

> With this configuration, we can run Trace-based tests even with a sampling into the main data store, having the following structure:

Reviewer: ```suggestion
With this configuration, we can run Trace-based tests, even with a sampling, into the main data store, having the following structure:
```

```mermaid
---
title: Observability architecture with Trace-based testing support
---
flowchart LR
    GoAPI["Go API"]
    OTelCol["OTel Collector"]
    DataStore[("Jaeger")]
    Tracetest

    GoAPI -- Send traces --> OTelCol
    OTelCol -- Forward 50% of Traces --> DataStore
    OTelCol -- Forward only `tracetest=true` traces --> Tracetest
``````

## Running the example

> If you want to run this example, just execute `docker compose up` on this folder.

Reviewer: ```suggestion
If you want to run this example, execute `docker compose up` on this folder.
```

> To execute a Trace-based test with Tracetest against this structure just run `tracetest run test -f test-api-working.yaml`.

Reviewer: ```suggestion
To execute a Trace-based test with Tracetest against this structure, run `tracetest run test -f test-api-working.yaml`.
```
